structmemfunclowering: fix crash when gep is optimized away

Hyxogen opened the pull request:
LLVM is smart enough to be able to remove a gep if the source and indexes are constants. However, it was assumed that it would always emit a gep instruction, which would cause a crash when trying to cast it.

Upon further inspection, I believe that this casting to a GEPOperator and then getting the result element type is overkill, and we could just simply use the type of the value.

This fixes: https://github.com/leaningtech/cheerp-meta/issues/141

yuri91 commented:
llvm is removing getPointerElementType(), so maybe that's why we do the overkill cast there. At some point we started removing all the "easy" cases of getPointerElementType usage.

yuri91 commented:
What is the actual type that we are trying to cast?

Hyxogen commented:
In the example from issue from cheerp-meta, it was `i16** null`

yuri91 commented:
Ah I see that it's a completely degenerate case. Can we just skip applying the pass in this case?

Hyxogen commented:
I'll check, I also saw this commit which I believe was never merged, it might also solve it:
https://github.com/leaningtech/cheerp-compiler/commit/9e9e5272b0624890a944480e1eb58a0a0afc73a2

yuri91 commented:
or even assuming that the type is char* if it's not inferrable from a gep

Hyxogen commented:
I've removed the use of `getPointerElementType`. Also upon further inspection, I believe that the surrounding if statement was
meant as a way of working around one of such optimizations. The new code should work for the general case, so I've removed it.

Hyxogen commented:
@yuri91

yuri91 reviewed (approved):
If the original issue is fixed, close it saying that it's fixed in master
